src/dsp/rde.c

```diff
@@ -1223,3 +1223,272 @@ int decode_rde_multipart_receive_resp(
+	uint8_t operation_type, bitfield8_t *operation_flags,
+	uint32_t send_data_transfer_handle, uint8_t operation_locator_length,
+	uint32_t request_payload_length, const uint8_t *operation_locator,
+	const uint8_t *request_payload, struct pldm_msg *msg)
+{
+	PLDM_MSGBUF_DEFINE_P(buf);
+	int rc;
+
+	if ((msg == NULL) || (operation_flags == NULL)) {
+		return PLDM_ERROR_INVALID_DATA;
+	}
+	if ((operation_locator_length > 0) && (operation_locator == NULL)) {
+		return PLDM_ERROR_INVALID_DATA;
+	}
+	if ((request_payload_length > 0) && (request_payload == NULL)) {
+		return PLDM_ERROR_INVALID_DATA;
+	}
+
+	struct pldm_header_info header;
+	header.instance = instance_id;
+	header.pldm_type = PLDM_RDE;
+	header.msg_type = PLDM_REQUEST;
+	header.command = PLDM_RDE_OPERATION_INIT;
+	rc = pack_pldm_header(&header, &(msg->hdr));
+
+	if (rc != PLDM_SUCCESS) {
+		return rc;
+	}
+
+	rc = pldm_msgbuf_init_errno(
+		buf, PLDM_RDE_OPERATION_INIT_REQ_FIXED_BYTES, msg->payload,
+		PLDM_RDE_OPERATION_INIT_REQ_FIXED_BYTES +
+			operation_locator_length + request_payload_length);
+
+	if (rc) {
+		return rc;
+	}
+
+	pldm_msgbuf_insert(buf, resource_id);
+	pldm_msgbuf_insert(buf, operation_id);
+	pldm_msgbuf_insert(buf, operation_type);
+	pldm_msgbuf_insert(buf, operation_flags->byte);
+	pldm_msgbuf_insert(buf, send_data_transfer_handle);
+	pldm_msgbuf_insert(buf, operation_locator_length);
+	pldm_msgbuf_insert(buf, request_payload_length);
+
+	if (operation_locator_length > 0) {
+		rc = pldm_msgbuf_insert_array(
+			buf, operation_locator_length,
+			(const uint8_t *)operation_locator,
+			operation_locator_length);
+		if (rc != PLDM_SUCCESS) {
+			return rc;
+		}
+	}
+
+	if (request_payload_length > 0) {
+		rc = pldm_msgbuf_insert_array(buf, request_payload_length,
+					      (const uint8_t *)request_payload,
+					      request_payload_length);
+		if (rc != PLDM_SUCCESS) {
+			return rc;
+		}
+	}
+
+	return pldm_msgbuf_complete(buf);
+}
+
+LIBPLDM_ABI_STABLE
+int decode_rde_operation_init_req(
+	const struct pldm_msg *msg, size_t payload_length,
+	uint32_t *resource_id, rde_op_id *operation_id, uint8_t *operation_type,
+	bitfield8_t *operation_flags, uint32_t *send_data_transfer_handle,
+	uint8_t *operation_locator_length, uint32_t *request_payload_length,
+	uint8_t *operation_locator, uint8_t *request_payload)
+{
+	PLDM_MSGBUF_DEFINE_P(buf);
+	int rc;
+
+	if ((msg == NULL) || (resource_id == NULL) || (operation_id == NULL) ||
+	    (operation_type == NULL) || (operation_flags == NULL) ||
+	    (send_data_transfer_handle == NULL) ||
+	    (operation_locator_length == NULL) ||
+	    (request_payload_length == NULL)) {
+		return PLDM_ERROR_INVALID_DATA;
+	}
+	if (payload_length < PLDM_RDE_OPERATION_INIT_REQ_FIXED_BYTES) {
+		return PLDM_ERROR_INVALID_LENGTH;
+	}
+
+	rc = pldm_msgbuf_init_errno(buf,
+				    PLDM_RDE_MULTIPART_SEND_REQ_FIXED_BYTES,
+				    msg->payload, payload_length);
+
+	if (rc) {
+		return rc;
+	}
+
+	pldm_msgbuf_extract_p(buf, resource_id);
```

Comment:
Need to extract the data , based on the spec defintions , optional data should default  Like Read/update fileds usage is dieffrent.